src/lib.rs

```diff
@@ -89,6 +89,59 @@ impl<K: std::hash::Hash + PartialEq, V> HashTable<K, V> {
     }
 }
 
+pub struct HashTableIterator<'a, K, V> {
+    current_bucket: Option<&'a Vec<(K, V)>>,
+    elements_iterator: Box<dyn Iterator<Item = &'a (K, V)> + 'a>,
+    buckets_iterator: Box<dyn Iterator<Item = &'a Vec<(K, V)>> + 'a>,
+}
+
+impl<'a, K, V> IntoIterator for &'a HashTable<K, V> {
+    type Item = &'a (K, V);
+
+    type IntoIter = HashTableIterator<'a, K, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let mut buckets_iterator = self.buckets.iter();
+        let current_bucket = buckets_iterator.next();
+        // first elements iterator needs to be initialized
+        let elements_iterator = match current_bucket {
+            Some(b) => b.iter(),
+            None => [].iter(),
+        };
+        HashTableIterator {
+            current_bucket,
+            elements_iterator: Box::new(elements_iterator),
+            buckets_iterator: Box::new(buckets_iterator),
+        }
+    }
+}
+
+impl<'a, K, V> Iterator for HashTableIterator<'a, K, V> {
+    type Item = &'a (K, V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // keep going on this bucket
+        // println!("keep going on this bucket");
+        self.elements_iterator.next().or_else(|| {
+            // println!("no element available in this bucket; iterating to next bucket and recursing");
+            self.current_bucket = self.buckets_iterator.next();
+            let b = match &self.current_bucket {
+                Some(b) => {
+                    // println!("bucket is avaiable");
+                    b
+                }
+                None => {
+                    // println!("no bucket is avaiable");
+                    return None;
```

Comment:
How else could I detect the end of the buckets iterator without this `match` and an explicit `return` statement?